crates/runtime/src/instance/allocator/pooling/memory_pool.rs

```diff
@@ -457,4 +749,143 @@ mod tests {
+        #[cfg_attr(miri, ignore)]
+        fn check_random_layout_calculations(c in constraints()) {
+            if let Ok(l) = calculate(&c) {
+                assert_slab_layout_invariants(c, l);
+            }
+        }
+    }
+
+    fn constraints() -> impl Strategy<Value = SlabConstraints> {
+        (
+            any::<usize>(),
+            any::<usize>(),
+            any::<usize>(),
+            any::<usize>(),
+            any::<bool>(),
+        )
+            .prop_map(
+                |(
+                    max_memory_bytes,
+                    num_memory_slots,
+                    num_pkeys_available,
+                    guard_bytes,
+                    guard_before_slots,
+                )| {
+                    SlabConstraints {
+                        max_memory_bytes,
+                        num_memory_slots,
+                        num_pkeys_available,
+                        guard_bytes,
+                        guard_before_slots,
+                    }
+                },
+            )
+    }
+
+    fn assert_slab_layout_invariants(c: SlabConstraints, s: SlabLayout) {
+        // Check that all the sizes add up.
+        assert_eq!(
+            s.total_slab_bytes,
+            s.pre_slab_guard_bytes + s.slot_bytes * c.num_memory_slots + s.post_slab_guard_bytes,
+            "the slab size does not add up: {c:?} => {s:?}"
+        );
+
+        // Check that the memory slot size is page-aligned.
+        assert!(
+            s.slot_bytes % crate::page_size() == 0,
+            "slot size is not page-aligned: {c:?} => {s:?}",
+        );
+
+        // Check that we use no more or less stripes than needed.
+        assert!(s.num_stripes >= 1, "not enough stripes: {c:?} => {s:?}");
+        if c.num_pkeys_available == 0 || c.num_memory_slots == 0 {
+            assert_eq!(
+                s.num_stripes, 1,
+                "expected at least one stripe: {c:?} => {s:?}"
+            );
+        } else {
+            assert!(
+                s.num_stripes <= c.num_pkeys_available,
+                "layout has more stripes than available pkeys: {c:?} => {s:?}"
+            );
+            assert!(
+                s.num_stripes <= c.num_memory_slots,
+                "layout has more stripes than memory slots: {c:?} => {s:?}"
+            );
+        }
+
+        // Check that we use the minimum number of stripes/protection keys.
+        // - if the next slot is bigger
+        if c.num_pkeys_available > 1 && c.max_memory_bytes > 0 {
+            assert!(
+                s.num_stripes <= (c.guard_bytes / c.max_memory_bytes) + 2,
+                "calculated more stripes than needed: {c:?} => {s:?}"
+            );
+        }
+
+        // Check that the memory-striping will not allow OOB access.
+        if s.num_stripes > 1 {
+            assert!(
+                s.slot_bytes * (s.num_stripes - 1) >= c.guard_bytes,
+                "layout may allow OOB access: {c:?} => {s:?}"
+            );
+        }
```

Comment:
This is one check which I think that we'll want to change. By taking into account the two variables I mentioned above I think this may need to change. 

Rather I think the rough invariant we want to test here is that for all slots in the layout there are guard bytes before, there's `static_memory_bound + guard` bytes after it to the next slot in the stripe. This I think may be best modeled with extra methods on the layout such as `fn slot_offset(&self, stripe: usize, slot: usize) -> usize` or something like that. This loop could then check:

```rust
for stripe in 0..stripes {
  for slot in 0..slots {
    assert_slot_is_valid(&layout, stripe, slot);
  }
}
```

more-or-less

Reply:
I realized as I went back over this that we don't ever need to iterate over slots: the layout has the sizes for every slot, so we just need to check that the layout is correct once: `assert!(s.slot_bytes * s.num_stripes >= c.expected_slot_bytes + c.guard_bytes)`.